src/librbd/io/Dispatcher.h

```diff
@@ -101,59 +133,81 @@ class Dispatcher : public DispatchInterfaceT {
+    if (dispatch_spec->dispatch_result == DISPATCH_RESULT_REFRESH) {
+      ldout(cct, 20) << "rstart. dispatch_spec=" << dispatch_spec << dendl;
+      m_op_tracker->start_op();
+    }
+
     // apply the IO request to all layers -- this method will be re-invoked
     // by the dispatch layer if continuing / restarting the IO
     while (true) {
-      m_lock.lock_shared();
       dispatch_layer = dispatch_spec->dispatch_layer;
       auto it = m_dispatches.upper_bound(dispatch_layer);
       if (it == m_dispatches.end()) {
         // the request is complete if handled by all layers
         dispatch_spec->dispatch_result = DISPATCH_RESULT_COMPLETE;
-        m_lock.unlock_shared();
         break;
       }
 
       auto& dispatch_meta = it->second;
       auto dispatch = dispatch_meta.dispatch;
-      auto async_op_tracker = dispatch_meta.async_op_tracker;
       dispatch_spec->dispatch_result = DISPATCH_RESULT_INVALID;
 
-      // prevent recursive locking back into the dispatcher while handling IO
-      async_op_tracker->start_op();
-      m_lock.unlock_shared();
-
       // advance to next layer in case we skip or continue
       dispatch_spec->dispatch_layer = dispatch->get_dispatch_layer();
 
       bool handled = send_dispatch(dispatch, dispatch_spec);
-      async_op_tracker->finish_op();
+      // This will end up in ImageDispatchSpec::C_Dispatcher::complete() where it will look at the dispatch_spec result.
+      // If it is CONTINUE then it will call back into this function (for the next layer)
+      // If it is COMPLETE then it will call finish - in ImageDispatchSpec this will just delete the dispatchSpec
+
+      // Alternatively this will end up in ObjectDispatchSpec::C_Dispatcher::complete() where it will look at the dispatch_spec result.
+      // If it is CONTINUE then it will call back into this function (for the next layer)
+      // If it is COMPLETE then it will call finish - this will call the complete() method (and then delete the object dispatchSpec)
+      //                                              the complete() method will call AioCompletion::complete which in turn will call AioCompletion::notify_callbacks_complete()
+      //                                              this will call ImageDispatchSpec::C_Dispatcher::complete() which will delete the image dispatchSpec
 
       // handled ops will resume when the dispatch ctx is invoked
       if (handled) {
-        return;
-      }
+
+        // If this IO request was queued waiting for a refresh to complete then decrement the count of outstanding operations.
+        // Once the refresh has completed the request will be resubmitted and tracked again.
+        // This is required because if the refresh needs to add or remove dispatch layers then it can only do so when there are no 
+        // IOs in flight.  IOs that are queued waiting for the refresh to complete are not in flight.
+        if(dispatch_spec->dispatch_result == DISPATCH_RESULT_REFRESH) {
```

Comment:
It isn't safe to access dispatch_spec here - there is a race where another thread has received the complete callback and has progressed dispatch_spec before this thread gets here. You need to be doing the decrement in need_refresh